File: lang/python/material/union_find.py
class unionFind:
    """
    Union-Find
    頂点を結合した際に連結する要素の親ノードを1つに統一する
    各頂点の親ノードを高速に返す
    """
    def __init__(self,n:int):
        """
        コンストラクタ

        引数:
            n (int):すべて独立したn個ノードとして初期化する
        """
        self.__par=[-1]*n

    def root(self,i:int)->int:
        """親ノードを再帰的に探索する
        途中で見つけた親ノードで各子ノード更新する
        引数:
            i (int):探したいノード 

        戻り値:
            int:親ノード
        """
        if self.__par[i]==-1:
            return i
        else:
            self.__par[i]=self.root(self.__par[i])
            return self.__par[i]

    def unite(self,a:int,b:int)->None:
        """ノードaとノードbを結合する

        引数:
            a (int): 結合したいノード
            b (int): 結合したいノード
        """
        b,a=self.root(b),self.root(a)
        if a==b:
            return
        self.__par[b]=a

File: dev/UnionFind.cpp
#include <vector>
#include <cstdint>
/**
 * @class UnionFind
 * @brief UnionFindの木を管理するクラス
 */
class UnionFind
{
public:
    /**
     * @brief unionFindのコンストラクタ
     * 初期化を行う
     * @param length 長さ
     */
    UnionFind(int64_t length) : m_groupLeaders(length, ROOT), m_sizeList(length, 1), m_rootCount(length)
    {
    }

    /**
     * @brief nodeAにnodeBを結合する
     * @param nodeA 結合対象のノード番号
     * @param nodeB 結合対象のノード番号
     */
    void Unite(const int64_t nodeA, const int64_t nodeB)
    {
        int64_t aRoot = UpdateRoot(nodeA);
        int64_t bRoot = UpdateRoot(nodeB);
        if (aRoot == bRoot)
        {
            return;
        }
        int64_t aSize = GetGroupSize(aRoot);
        int64_t bSize = GetGroupSize(bRoot);
        if (aSize < bSize)
        {
            std::swap(aRoot, bRoot);
            std::swap(aSize, bSize);
        }
        m_sizeList[aRoot] += bSize;
        m_sizeList[bRoot] = 0;
        m_groupLeaders[bRoot] = aRoot;
        --m_rootCount;
    }

    /**
     * @brief ノードの属するグループのサイズを求める
     * @param node ノード番号
     */
    int64_t GetGroupSize(const int64_t node)
    {
        return m_sizeList[UpdateRoot(node)];
    }
    /**
     * 根の数を数える
     */
    int64_t GetRootCount() const
    {
        return m_rootCount;
    }

    /**
     * @brief ノードが根かどうかを返す
     * @param node ノード番号
     * @return ノードが根かどうか
     */
    bool IsRoot(const int64_t node) const
    {
        return m_groupLeaders[node] == ROOT;
    }

    /**
     * @brief ノード同士が同じグループかどうかを求める
     * @param nodeA 調べる対象のノード
     * @param nodeB 調べる対象のノード
     * @return 同じグループかどうか
     */
    bool IsSameRoot(const int64_t nodeA, const int64_t nodeB)
    {
        return GetRoot(nodeA) == GetRoot(nodeB);
    }

    /**
     * @brief nodeの根を更新し、根を取得する
     * @param 調べたい対象のノード番号
     */
    int64_t GetRoot(const int64_t node)
    {
        return UpdateRoot(node);
    }

private:
    /**
     * @brief nodeの根を更新する
     * また、根に至るまでの経路をすべて根の子とする
     * @param node 更新対象のノード番号
     * @return 根のノード番号
     */
    int64_t UpdateRoot(const int64_t node)
    {
        if (IsRoot(node))
        {
            return node;
        }
        m_groupLeaders[node] = UpdateRoot(m_groupLeaders[node]);
        return m_groupLeaders[node];
    }

    std::vector<int64_t> m_groupLeaders; //! 自分の親ノードが誰かの情報を持つ隣接リストvector
    std::vector<int64_t> m_sizeList;     //! そのグループのサイズを持つvector
    int64_t m_rootCount;                 //! 根の数を持つ値
    static constexpr int ROOT = -1;      //! そのノードが根であることを示す値
};
